R/spatial_block_cv.R

```diff
@@ -104,54 +118,67 @@ spatial_block_cv <- function(data, method = "random", v = 10, ...) {
     splits = split_objs$splits,
     ids = split_objs[, grepl("^id", names(split_objs))],
     attrib = cv_att,
-    subclass = c("spatial_block_cv", "rset")
+    subclass = c("spatial_block_cv", "spatial_rset", "rset")
   )
 
 }
 
 random_block_cv <- function(data, grid_blocks, v) {
   n <- nrow(data)
 
-  block_contains_points <- purrr::map_lgl(
-    sf::st_intersects(grid_blocks, data),
-    sgbp_is_not_empty
-  )
-  grid_blocks <- grid_blocks[block_contains_points]
+  grid_blocks <- filter_grid_blocks(grid_blocks, data)
 
   n_blocks <- length(grid_blocks)
-  if (!is.numeric(v) || length(v) != 1) {
-    rlang::abort("`v` must be a single integer.")
-  }
-  if (v > n_blocks) {
-    rlang::warn(paste0(
-      "Fewer than ", v, " blocks available for sampling; setting v to ",
-      n_blocks, "."
-    ))
-    v <- n_blocks
-  }
+  v <- check_v(v, n_blocks, "blocks")
 
   grid_blocks <- sf::st_as_sf(grid_blocks)
   grid_blocks$fold <- sample(rep(seq_len(v), length.out = nrow(grid_blocks)))
-  grid_blocks <- split_unnamed(grid_blocks, grid_blocks$fold)
 
-  # grid_blocks is now a list of sgbp lists (?sf::sgbp)
-  #
-  # The first map() here iterates through the meta-list,
-  # and the second checks each element of the relevant sgbp list
-  # to see if it is integer(0) (no intersections) or not
-  #
-  # Each sgbp sub-list is nrow(data) elements long, so this which()
-  # returns the list indices which are not empty, which is equivalent
-  # to the row numbers that intersect with blocks in the fold
-  indices <- purrr::map(
-    grid_blocks,
-    function(blocks) which(
-      purrr::map_lgl(
-        sf::st_intersects(data, blocks),
-        sgbp_is_not_empty
+  generate_folds_from_blocks(data, grid_blocks, v, n)
+}
+
+systematic_block_cv <- function(data, grid_blocks, v,
+                                ordering = c("snake", "continuous"),
+                                relevant_only = TRUE) {
+  n <- nrow(data)
+  ordering <- rlang::arg_match(ordering)
+
+  if (relevant_only) grid_blocks <- filter_grid_blocks(grid_blocks, data)
+
+  n_blocks <- length(grid_blocks)
+  v <- check_v(v, n_blocks, "blocks")
+
+  folds <- rep(seq_len(v), length.out = length(grid_blocks))
+  if (ordering == "snake") folds <- make_snake_ordering(folds, grid_blocks)
+
+  grid_blocks <- sf::st_as_sf(grid_blocks)
+  grid_blocks$fold <- folds
+  if (!relevant_only) grid_blocks <- filter_grid_blocks(grid_blocks, data)
+
+  num_folds <- length(unique(grid_blocks$fold))
+  if (num_folds != v) {
+    rlang::warn(
+      c(
+        paste0(
+          "Not all folds contained blocks with data: \n",
+          v,
+          " folds were requested, but only ",
+          num_folds,
+          " contain any data. \nEmpty folds were dropped."
+        ),
```

Comment:
Let's switch over to `glue::glue()` for our error messages in this package, starting here. You can add glue to Imports.